README
Read Me(ucsd-psystem-xc)                              Read Me(ucsd-psystem-xc)



NAME
       ucsd-psystem-xc - UCSD p-System Pascal cross compiler

DESCRIPTION
       The ucsd-psystem-xc package is a collection of tools for compiling
       Pascal source files to produce UCSD p_System code files.  The package
       includes:

       ucsdpsys(1)
               A laucher to run the virtual machine comfortably from the
               command line.  It includes a batch mode for automating
               (scripting) operations.

       ucsdpsys_assemble(1)
               The cross assembler.  It is able to assemble several different
               target microprocessor architectures in the one executable.

       ucsdpsys_compile(1)
               The cross compiler.  It understands the UCSD Pascal dialect,
               including UNIT definitions and references.

       ucsdpsys_depends(1)
               May be used to determine include file dependencies, for use
               with make(1) and other build tools.

       ucsdpsys_disassemble(1)
               For disassembling UCSD p-System code files.  This is used to
               verify the correctness of the compiler.

       ucsdpsys_downcase(1)
               A untility for converting Pascal code to lower case, leaving
               string constants and comments unaltered.

       ucsdpsys_errors(1)
               A utility to translate back and forth between text and binary
               representations of the assembler error message files.

       ucsdpsys_libmap(1)
               A utility for printing segment maps of UCSD p-System library
               files.

       ucsdpsys_librarian(1)
               A utility for manipulating the segments within UCSD p-System
               codefiles.

       ucsdpsys_link(1)
               A utility for linking UCSD p-System codefiles to their
               assembler components.

       ucsdpsys_opcodes(1)
               A utility to translate back and forth between text and binary
               representations of the assembler opcode files.

       ucsdpsys_setup(1)
               A utility to translate back and forth between text and binary
               representations of the system.miscinfo file.

   Sister Projects
       Some other projects will be of interest to you.

       ucsd-psystem-fs
               This package contains tools for manipulating UCSD p-System
               floppy disk images, and a file system for mounting them in
               Linux as real file systems.
               http://ucsd-psystem-fs.sourceforge.net/

       ucsd-psystem-os
               This project provides a self-hosting set of system sources.
               You need the disk images produced by this project for the
               virtual machine to have a "system.pascal" file to run (this
               provides runtime support and the user command executive).  This
               is a work in progress.

       uvsd-psystem-vm
               This package provides a fully featured UCSD p-Machine emulator.

ARCHIVE SITE
       The latest version of ucsd-psystem-xc is available on the Web from:

           URL:    http://ucsd-psystem-xc.sourceforge.net/
           File:   ucsd-psystem-xc-0.13.README   # Description, from the tar file
           File:   ucsd-psystem-xc-0.13.lsm      # Description, LSM format
           File:   ucsd-psystem-xc-0.13.tar.gz   # the complete source
           File:   ucsd-psystem-xc-0.13.pdf      # Reference Manual

BUILDING ucsd-psystem-xc
       Full instructions for building ucsd-psystem-xc may be found in the
       BUILDING file included in this distribution.

COPYRIGHT
       ucsd-psystem-xc version 0.13
       Copyright (C) 2006, 2007, 2010, 2011, 2012 Peter Miller

       This program is free software; you can redistribute it and/or modify it
       under the terms of the GNU General Public License as published by the
       Free Software Foundation; either version 3 of the License, or (at your
       option) any later version.

       This program is distributed in the hope that it will be useful, but
       WITHOUT ANY WARRANTY; without even the implied warranty of
       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
       General Public License for more details.

       You should have received a copy of the GNU General Public License along
       with this program. If not, see <http://www.gnu.org/licenses/>.

       It should be in the LICENSE file included with this distribution.

AUTHOR
       Peter Miller   E-Mail:   pmiller@opensource.org.au
       /\/\*          WWW:      http://miller.emu.id.au/pmiller/

RELEASE NOTES
       This section details the features and bug fixes of each of the
       releases.

   Version 0.13 (2012-Nov-12)
       * The ucsdpsys_charset(1) command now understands more font file
         formats, including Terak and PSF Tools format.

       * The ucsdpsys_charset(1) command has a new -negative option, that is
         used to calculate the inverted "top half" of a font from the normal
         "bottom half", as is common in Terak fonts.

       * The ucsdpsys_charset(1) is now able manipulate the boot logo in Terak
         system.charset files.

       * The ucsdpsys_osmakgen(1) command now takes advantage of recent
         ucsdpsys_osmakgen(1) features.

       * There is a new ucsdpsys_foto(1) command, used to convert UCSD Pascal
         .foto files into .png files, and vice versa.

   Version 0.12 (2012-Nov-02)
       * The ucsdpsys_osmakgen(1) command now understands how to process
         SYSTEM.CHARSET files.

       * There is a new ucsdpsys_osmakgen(1) option called --arch-from-host,
         that may be used to translate a host name (e.g. "Terak") into an arch
         name (e.g.  "pdp11").

       * The ucsdpsys_charset(1) command now understands a
         --architecture=terak option, which means to work on a SYSTEM.CHARSET
         file suitable for a Terak system, where the glyphs are 8x10 and laid
         out differently in the binary file.

   Version 0.11 (2012-Jul-28)
       * Kai Henningsen <kai.extern@gmail.com> discovered that 'Makefile'
         files generated by ucsdpsys_osmakgen did not correctly support the
         'distclean' target. This has been corrected.

       * Work is in progress to be able to cope with multiple p-machine
         versions.

       * The compiler is now able to cope with variables declared in plain
         units.

       * The ucsdpsys_osmakgen(1) command now understands how to generate the
         necessary debian/ files for building a debian package from the ucsd
         p-system operating system sources.

       * Thw ucsdpsys(1) file no longer creates the implied system disk image
         if one of the supplied disk images is a functioning system disk.

       * The ucsdpsys(1) command now better understands where ucsd-psystem-os
         installs its files, which it needs in order to build the default
         system disk image.

   Version 0.10 (2011-May-18)
       * A bug which caused a segfault in the ubsdpsys --batch option has been
         fixed.

       * The ucsdpsys_osmakgen(1) command, used by the ucsd-psystem-os project
         to generate its Makefile, now understands the presence of man pages,
         and installs them appropriately.

   Version 0.9 (2011-Feb-02)
       * The slides of the LCA 2011 talk "Factory Factory Factories" is now
         available in the web site.

       * The ucsdpsys_osmakgen(1) command has been improved, with a view to
         Debian packaging of the OS.

       * The ucsdpsys(1) command has a new --no-system option, to suppress the
         construction of a system disk image.

       * There is a new ucsdpsys_compile(1) option, --library-path for adding
         directories to the library search path.

       * The ucsdpsys_compile(1) command now fully supports the (*$U filename
         *) control comment.

       * The ucsdpsys_assemble(1) command now understands the .error .print
         .sbttl .title pseudo-ops, mostly named for PDP-11 assmebler pseudo-
         ops of the same name.

       * The ucsdpsys_charset(1) command has been moved to this project, out
         of the ucsd-psystem-fs project.

       * The ucsdpsys_assemble(1) command now understands how to produce
         assembler listings, using the -L option.  See ucsdpsys_assemble(1)
         for more information.

       * The ucsdpsys_compile(1) command now issues warnings for unreachange
         statements. There is a new (*$warning unreachable false *) control
         comment to disable the warning.

       * The project download web page now includes a link to the LunchPad
         PPA, where pre-compiled Ubuntu packages are available.

       * The ucsdpsys_assemble(1) command now understands the .ref pseudo-op,
         and generates the appropriate relocation information.

       * The ucsdpsys_assemble(1) command now more closely emulates the UCSD
         native assembler, in the way it forgets symbols created between one
         .proc and another.  This stops historical source files from
         complaining about multiply defined symbols all over the place.

       * The ucsdpsys_assemble(1) command now requires that the architecture
         be explicitly stated, either with the .arch pseudo-op, or the --arch
         command line option, in all cases.

       * The ucsdpsys_assemble(1) command now ignores all input after the .end
         directive.

       * The ucsdpsys_assembler(1) command now understands .gt greater than,
         >= greater than or equal, .lt less than, <= less than or equal, <>
         inequality, and = equality comparisons.

       * The ucsdpsys_assemble(1) command, now understands, for 6502 opcodes,
         how to relocate segment relative addresses for absolute addressing
         opcodes.

       * The ucsdpsys_assemble(1) command now understands conditional assembly
         .if, .else and .endc pseudo-op directives.

       * The ucsdpsys_assembler(1) command now understands the .macro pseudo-
         op, for defining an substituting macros into the code stream.

       * A bug has been fixed in the code that checks codefiles for validity.
         It no longer rejects segment dictionaries with zero-length UNITSEG
         segments.  These are produced when a program USES a non-intrinsic
         unit, but is not yet linked.

   Version 0.8 (2010-Aug-28)
       * The ucsdpsys_assemble(1) cross assembler now understands the .func
         pseudo-op.

       * The error message formatting has been changed to use a 4 character
         hanging indent for multi-line error messages.

       * A bug has been fixed in the ucsdpsys_osmakgen(1) command, it now
         correctly understands how to remove system segments from libraries
         with an assembler component.

       * The ucsdpsys_osmakgen(1) command now understands how to link Pascal
         programs with their assembler components.

       * A bug has been fixed in the ucsdpsys(1) command, it no longer fails
         if its temporary files are unlinked twice.

       * There is a new ucsdpsys_compile(1) --view-path option, symmetric with
         the ucsdpsys_assemble(1) and ucsdpsys_depends(1) commands' options of
         the same name.

       * The ucsdpsys_assemble(1) command now understands the .incude pseudo-
         op.  This is also a new corresponding -I command line option.

       * A bug has been fixed in the ucsdpsys_librarian(1) command, it now
         patches the segment number in the procedure dictionary when it
         renumbers a segment.

       * A bug has been fixed in the ucsdpsys_disassemble(1) and
         ucsdpsys_libmap(1) commands, they were printing SEPPROC link
         information incorrectly.

       * The ucsdpsys_osmakgen(1) command now generates an "install" target,
         so that the results of the build can be installed into the system.

       * The ucsdpsys_assemble(1) cross assembler now groks unary minus (-e)
         unary plus (+e) bit-wise and (e1 & e2), bit-wise or (e1 | e2), bit-
         wise not (~e), bit-wise exclusive-or (e1 ^ e2), and modulo (e1 % e2)

debian/changelog
ucsd-psystem-xc (0.13.D001-1) unstable; urgency=low

  * This change set prepares the ucsd-psystem-xc project for the next public
    release.

 -- Peter Miller <pmiller@opensource.org.au>  Mon, 12 Nov 2012 13:11:38 +1100

ucsd-psystem-xc (0.12.D034-1) unstable; urgency=low

  * This change set takes heed of the ASpple Pascla documentation for
    SYSTEM.CHARSET and treats glyphs as 7x8 whenever possible.

 -- Peter Miller <pmiller@opensource.org.au>  Mon, 12 Nov 2012 11:21:34 +1100

ucsd-psystem-xc (0.12.D033-1) unstable; urgency=low

  * The ucsdpsys_charset(1) command has a new --negative option, which is used
    to calculate the inverted 'top half' from the normal 'bottom half', as is
    common in Terak fonts.

 -- Peter Miller <pmiller@opensource.org.au>  Mon, 12 Nov 2012 10:51:24 +1100

ucsd-psystem-xc (0.12.D032-1) unstable; urgency=low

  * The ucsdpsys_charset(1) command is now able to write font defintiions using
    the PSF Tools text format.	Using PSF  Tools, you willbe able to import and
    exporrt many different font formats.

 -- Peter Miller <pmiller@opensource.org.au>  Sun, 11 Nov 2012 21:44:05 +1100

ucsd-psystem-xc (0.12.D031-1) unstable; urgency=low

  * The ucsdpsys_charset(1) command is now able to read PSF Tools formatted font
    definition text files.

 -- Peter Miller <pmiller@opensource.org.au>  Sun, 11 Nov 2012 19:56:03 +1100

ucsd-psystem-xc (0.12.D030-1) unstable; urgency=low

  * This change set adds vim-mode lines to th last 25% of files.

 -- Peter Miller <pmiller@opensource.org.au>  Sun, 11 Nov 2012 06:09:03 +1100

ucsd-psystem-xc (0.12.D029-1) unstable; urgency=low

  * This change set does the next 25% of source files and vim mode lines.

 -- Peter Miller <pmiller@opensource.org.au>  Sun, 11 Nov 2012 00:27:08 +1100

ucsd-psystem-xc (0.12.D028-1) unstable; urgency=low

  * This change set adds more vim mode lines, to the next 25%

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 10 Nov 2012 23:28:55 +1100

ucsd-psystem-xc (0.12.D027-1) unstable; urgency=low

  * This change set adds vim-mode lines to 24% of the source files

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 10 Nov 2012 22:45:07 +1100

ucsd-psystem-xc (0.12.D026-1) unstable; urgency=low

  * This change set adds the ability write a boot logo into the system.charset
    file.  (Just exactly *why* they put the boot logo in such an unlikely place
    is anyone's guess.)

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 10 Nov 2012 21:26:11 +1100

ucsd-psystem-xc (0.12.D025-1) unstable; urgency=low

  * This change set adds insulation for <getopt.h>

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 10 Nov 2012 15:28:14 +1100

ucsd-psystem-xc (0.12.D024-1) unstable; urgency=low

  * This change set breaks up the input and output classes, so that we can get
    squared functionality, with linear code.

 -- Peter Miller <pmiller@opensource.org.au>  Fri, 09 Nov 2012 19:20:31 +1100

ucsd-psystem-xc (0.12.D023-1) unstable; urgency=low

  * This change set breaks up the input and output classes, so that we can get
    squared functionality, with linear code.

 -- Peter Miller <pmiller@opensource.org.au>  Fri, 09 Nov 2012 17:39:59 +1100

ucsd-psystem-xc (0.12.D022-1) unstable; urgency=low

  * This change et improves the way ucsdpsys_osmakgen(1) deals with '.charset'
    files.

 -- Peter Miller <pmiller@opensource.org.au>  Wed, 07 Nov 2012 12:28:44 +1100

ucsd-psystem-xc (0.12.D021-1) unstable; urgency=low

  * This change set makes the ucsdpsys_charset --host option actually consider
    host names. It fixes a bunch of other commands, too, for the same thing.

 -- Peter Miller <pmiller@opensource.org.au>  Wed, 07 Nov 2012 11:42:00 +1100

ucsd-psystem-xc (0.12.D020-1) unstable; urgency=low

  * This change set fixes two doxygen warnings; the html needed some escaping

 -- Peter Miller <pmiller@opensource.org.au>  Wed, 07 Nov 2012 08:18:06 +1100

ucsd-psystem-xc (0.12.D019-1) unstable; urgency=low

  * This change set adds a link to PSF-Tools in the ucsdpsys_charset(1) man
    page.

 -- Peter Miller <pmiller@opensource.org.au>  Mon, 05 Nov 2012 12:02:25 +1100

ucsd-psystem-xc (0.12.D018-1) unstable; urgency=low

  * This change set adds text to ucsdpsys_charset(1) man page to document the
    formats of the system.charset files.

 -- Peter Miller <pmiller@opensource.org.au>  Mon, 05 Nov 2012 03:48:55 +1100

ucsd-psystem-xc (0.12.D017-1) unstable; urgency=low

  * There is a new ucsdpsys_foto(1) command, used to convert ucsd pascal .foto
    files into .png files, and back again.

 -- Peter Miller <pmiller@opensource.org.au>  Sun, 04 Nov 2012 23:51:45 +1100

ucsd-psystem-xc (0.12.D016-1) unstable; urgency=low

  * This change set confirms that the Terak Foto Display program works.	 All
    works as expected.	It did show up an error case, where an array was *way*
    too big.

 -- Peter Miller <pmiller@opensource.org.au>  Sun, 04 Nov 2012 17:43:28 +1100

ucsd-psystem-xc (0.12.D015-1) unstable; urgency=low

  * This change set fixes a build problem. My thanks to theLaunchPad PPA build
    farm for finding this problem.

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 03 Nov 2012 23:55:45 +1100

ucsd-psystem-xc (0.12.D014-1) unstable; urgency=low

  * This change set adds some C++ include insulation, covering <float.h>
    <limits.h> <math.h> <signal.h>

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 03 Nov 2012 21:09:17 +1100

ucsd-psystem-xc (0.12.D013-1) unstable; urgency=low

  * This change set adds more include insulation, this time for <iostream> and
    <sys/stat.h>

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 03 Nov 2012 20:26:32 +1100

ucsd-psystem-xc (0.12.D012-1) unstable; urgency=low

  * This change set adds more insulation fo rthe <cstdlib> and <ctime> headers.

 -- Peter Miller <pmiller@opensource.org.au>  Sat, 03 Nov 2012 19:41:44 +1100

ucsd-psystem-xc (0.12.D011-1) unstable; urgency=low

  * This change set adds insulation for <cctype> includes.


lib/pcode.h
//
// UCSD p-System cross compiler
// Copyright (C) 2006, 2007, 2010-2012 Peter Miller
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or (at
// you option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program. If not, see <http://www.gnu.org/licenses/>
//

#ifndef LIB_PCODE_H
#define LIB_PCODE_H

enum pcode_t
{
    SLDC_0,     // 00 Short Load Constant
    SLDC_1,     // 01 Short Load Constant
    SLDC_2,     // 02 Short Load Constant
    SLDC_3,     // 03 Short Load Constant
    SLDC_4,     // 04 Short Load Constant
    SLDC_5,     // 05 Short Load Constant
    SLDC_6,     // 06 Short Load Constant
    SLDC_7,     // 07 Short Load Constant
    SLDC_8,     // 08 Short Load Constant
    SLDC_9,     // 09 Short Load Constant
    SLDC_10,    // 0A Short Load Constant
    SLDC_11,    // 0B Short Load Constant
    SLDC_12,    // 0C Short Load Constant
    SLDC_13,    // 0D Short Load Constant
    SLDC_14,    // 0E Short Load Constant
    SLDC_15,    // 0F Short Load Constant
    SLDC_16,    // 10 Short Load Constant
    SLDC_17,    // 11 Short Load Constant
    SLDC_18,    // 12 Short Load Constant
    SLDC_19,    // 13 Short Load Constant
    SLDC_20,    // 14 Short Load Constant
    SLDC_21,    // 15 Short Load Constant
    SLDC_22,    // 16 Short Load Constant
    SLDC_23,    // 17 Short Load Constant
    SLDC_24,    // 18 Short Load Constant
    SLDC_25,    // 19 Short Load Constant
    SLDC_26,    // 1A Short Load Constant
    SLDC_27,    // 1B Short Load Constant
    SLDC_28,    // 1C Short Load Constant
    SLDC_29,    // 1D Short Load Constant
    SLDC_30,    // 1E Short Load Constant
    SLDC_31,    // 1F Short Load Constant
    SLDC_32,    // 20 Short Load Constant
    SLDC_33,    // 21 Short Load Constant
    SLDC_34,    // 22 Short Load Constant
    SLDC_35,    // 23 Short Load Constant
    SLDC_36,    // 24 Short Load Constant
    SLDC_37,    // 25 Short Load Constant
    SLDC_38,    // 26 Short Load Constant
    SLDC_39,    // 27 Short Load Constant
    SLDC_40,    // 28 Short Load Constant
    SLDC_41,    // 29 Short Load Constant
    SLDC_42,    // 2A Short Load Constant
    SLDC_43,    // 2B Short Load Constant
    SLDC_44,    // 2C Short Load Constant
    SLDC_45,    // 2D Short Load Constant
    SLDC_46,    // 2E Short Load Constant
    SLDC_47,    // 2F Short Load Constant
    SLDC_48,    // 30 Short Load Constant
    SLDC_49,    // 31 Short Load Constant
    SLDC_50,    // 32 Short Load Constant
    SLDC_51,    // 33 Short Load Constant
    SLDC_52,    // 34 Short Load Constant
    SLDC_53,    // 35 Short Load Constant
    SLDC_54,    // 36 Short Load Constant
    SLDC_55,    // 37 Short Load Constant
    SLDC_56,    // 38 Short Load Constant
    SLDC_57,    // 39 Short Load Constant
    SLDC_58,    // 3A Short Load Constant
    SLDC_59,    // 3B Short Load Constant
    SLDC_60,    // 3C Short Load Constant
    SLDC_61,    // 3D Short Load Constant
    SLDC_62,    // 3E Short Load Constant
    SLDC_63,    // 3F Short Load Constant
    SLDC_64,    // 40 Short Load Constant
    SLDC_65,    // 41 Short Load Constant
    SLDC_66,    // 42 Short Load Constant
    SLDC_67,    // 43 Short Load Constant
    SLDC_68,    // 44 Short Load Constant
    SLDC_69,    // 45 Short Load Constant
    SLDC_70,    // 46 Short Load Constant
    SLDC_71,    // 47 Short Load Constant
    SLDC_72,    // 48 Short Load Constant
    SLDC_73,    // 49 Short Load Constant
    SLDC_74,    // 4A Short Load Constant
    SLDC_75,    // 4B Short Load Constant
    SLDC_76,    // 4C Short Load Constant
    SLDC_77,    // 4D Short Load Constant
    SLDC_78,    // 4E Short Load Constant
    SLDC_79,    // 4F Short Load Constant
    SLDC_80,    // 50 Short Load Constant
    SLDC_81,    // 51 Short Load Constant
    SLDC_82,    // 52 Short Load Constant
    SLDC_83,    // 53 Short Load Constant
    SLDC_84,    // 54 Short Load Constant
    SLDC_85,    // 55 Short Load Constant
    SLDC_86,    // 56 Short Load Constant
    SLDC_87,    // 57 Short Load Constant
    SLDC_88,    // 58 Short Load Constant
    SLDC_89,    // 59 Short Load Constant
    SLDC_90,    // 5A Short Load Constant
    SLDC_91,    // 5B Short Load Constant
    SLDC_92,    // 5C Short Load Constant
    SLDC_93,    // 5D Short Load Constant
    SLDC_94,    // 5E Short Load Constant
    SLDC_95,    // 5F Short Load Constant
    SLDC_96,    // 60 Short Load Constant
    SLDC_97,    // 61 Short Load Constant
    SLDC_98,    // 62 Short Load Constant
    SLDC_99,    // 63 Short Load Constant
    SLDC_100,   // 64 Short Load Constant
    SLDC_101,   // 65 Short Load Constant
    SLDC_102,   // 66 Short Load Constant
    SLDC_103,   // 67 Short Load Constant
    SLDC_104,   // 68 Short Load Constant
    SLDC_105,   // 69 Short Load Constant
    SLDC_106,   // 6A Short Load Constant
    SLDC_107,   // 6B Short Load Constant
    SLDC_108,   // 6C Short Load Constant
    SLDC_109,   // 6D Short Load Constant
    SLDC_110,   // 6E Short Load Constant
    SLDC_111,   // 6F Short Load Constant
    SLDC_112,   // 70 Short Load Constant
    SLDC_113,   // 71 Short Load Constant
    SLDC_114,   // 72 Short Load Constant
    SLDC_115,   // 73 Short Load Constant
    SLDC_116,   // 74 Short Load Constant
    SLDC_117,   // 75 Short Load Constant
    SLDC_118,   // 76 Short Load Constant
    SLDC_119,   // 77 Short Load Constant
    SLDC_120,   // 78 Short Load Constant
    SLDC_121,   // 79 Short Load Constant
    SLDC_122,   // 7A Short Load Constant
    SLDC_123,   // 7B Short Load Constant
    SLDC_124,   // 7C Short Load Constant
    SLDC_125,   // 7D Short Load Constant
    SLDC_126,   // 7E Short Load Constant
    SLDC_127,   // 7F Short Load Constant
    ABI,        // 80 Absolute value integer
    ABR,        // 81 Absolute value real
    ADI,        // 82 Add integers
    ADR,        // 83 Add reals
    LAND,       // 84 Logical AND
    DIF,        // 85 Set difference
    DVI,        // 86 Divide integers
    DVR,        // 87 Divide reals
    CHK,        // 88 Check subrange bounds
    FLO,        // 89 Float top-of-stack-1
    FLT,        // 8A Float top-of-stack
    INN,        // 8B Test set membership
    INT,        // 8C Set intersection
    LOR,        // 8D Logical OR
    MODI,       // 8E Modulo integers
    MPI,        // 8F Multiply integers
    MPR,        // 90 Multiply reals
    NGI,        // 91 Negate integer
    NGR,        // 92 Negate real
    LNOT,       // 93 Logical NOT
    SRS,        // 94 Build subrange set
    SBI,        // 95 Subtract integers
    SBR,        // 96 Subtract reals
    SGS,        // 97 Build one-member set
    SQI,        // 98 Square integer
    SQR,        // 99 Square real
    STO,        // 9A Store Indirect
    IXS,        // 9B Index String array
    UNI,        // 9C Set union

    LDE,        // 9D II.1: Load Extended (byte, big)
                // 9D II.0: not imp
    S2P = LDE,  // 9D I.3, I.5: string to packed convert

    CSP,        // 9E I.5, II.0, II.1: Call standard procedure (byte)
                // 9E I.3: not imp

    LDCN,       // 9F Load constant NIL
    ADJ,        // A0 Adjust set (byte)
    FJP,        // A1 False jump (byte)
    INC,        // A2 Incriment field pointer (big)
    IND,        // A3 Static index and load (big)
    IXA,        // A4 Index array (big)
    LAO,        // A5 Load Global Address (big)

    LSA,        // A6 II.0, II.1: Load constant String Address
    LCA = LSA,  // A6 I.3, I.5: Load Constant string Address

    LAE,        // A7 II.1: Load Extended Address (byte, big)
    LDO_i_5 = LAE, // A7 I.3, I.5: LoaD Global word
               // A7 II.0: not imp

    MOV,        // A8 Move words (big)

    LDO,        // A9 II.0, II.1: Load Global (big)
    MVB = LDO,  // A9 I.3, I.5: MoVe Bytes

    SAS,        // AA String Assign
    SRO,        // AB Store Global (big)
    XJP,        // AC Case jump (word, word, ..., word)

    RNP,        // AD I.5, II.0, II.1: Return from non-base procedure (byte)
                // AD 1.3: not imp

    CIP,        // AE I.5, II.0, II.1: Call intermediate procedure (byte)
                // AE I.3: not imp

    EQU,        // AF Test equality (byte)
    GEQ,        // B0 test greater than or equal (byte)
    GTR,        // B1 Test greater than (byte)
    LDA,        // B2 Load Intermediate Address (big)
    LDC,        // B3 Load Multi-word Constant (byte, data...)
    LEQ,        // B4 Test less than or equal (byte)
    LES,        // B5 Test less than (byte)
    LOD,        // B6 Load Intermediate (big)
    NEQ,        // B7 Test inequality (byte)
    STR,        // B8 Stote Intermediate (byte, big)
    UJP,        // B9 Unsigned jump (byte)
    LDP,        // BA Load packed field
    STP,        // BB Store packed field
    LDM,        // BC Load Multi-word (byte)
    STM,        // BD Store Multi-word (byte)
    LDB,        // BE Load Byte
    STB,        // BF Store Byte
    IXP,        // C0 Index packed array (bytes, byte)
    RBP,        // C1 Return from base procedure (byte)
    CBP,        // C2 Call base procedure (byte)
    EQUI,       // C3 Test integer equality
    GEQI,       // C4 Test integer greater than or equal
    GTRI,       // C5 Test integer greater than
    LLA,        // C6 Load Local Address (big)
    LDCI,       // C7 Load constant integer (word)
    LEQI,       // C8 Test integer less than or equal
    LESI,       // C9 Test integer less than
    LDL,        // CA Load Local (big)
    NEQI,       // CB Test integer inequality
    STL,        // CC Store Local (big)
    CXP,        // CD Call external procedure (byte, byte)
    CLP,        // CE Call local procedure (byte)
    CGP,        // CF Call global procedure (byte)

    LPA,        // D0 II.0, II.1: Load packed array (byte, data...)
    S1P = LPA,  // D0 I.3, I.5: string to packed convert

    STE,        // D1 II.1: Store Extended (byte, big)
    IXB = STE,  // D1 I.3, I.5: IndeX Byte
                // D1 II.0: not imp

    BYT,        // D2 I.3, I.5: convert word pointer to byte pointer
                // D2 II.0, II.1: not imp

    EFJ,        // D3 Equal False Jump (byte)     => "branch if not equal"
    NFJ,        // D4 Not equal False Jump (byte) => "branch if equal"
    BPT,        // D5 Breakpoint (byte)
    XIT,        // D6 Exit the operating system.
    NOP,        // D7 No-operation
    SLDL_1,     // D8 Short Load Local
    SLDL_2,     // D9 Short Load Local
    SLDL_3,     // DA Short Load Local
    SLDL_4,     // DB Short Load Local
    SLDL_5,     // DC Short Load Local
    SLDL_6,     // DD Short Load Local
    SLDL_7,     // DE Short Load Local
    SLDL_8,     // DF Short Load Local
    SLDL_9,     // E0 Short Load Local
    SLDL_10,    // E1 Short Load Local
    SLDL_11,    // E2 Short Load Local
    SLDL_12,    // E3 Short Load Local
    SLDL_13,    // E4 Short Load Local
    SLDL_14,    // E5 Short Load Local
    SLDL_15,    // E6 Short Load Local
    SLDL_16,    // E7 Short Load Local
    SLDO_1,     // E8 Short Load Global
    SLDO_2,     // E9 Short Load Global
    SLDO_3,     // EA Short Load Global
    SLDO_4,     // EB Short Load Global
    SLDO_5,     // EC Short Load Global
    SLDO_6,     // ED Short Load Global
    SLDO_7,     // EE Short Load Global
    SLDO_8,     // EF Short Load Global
    SLDO_9,     // F0 Short Load Global
    SLDO_10,    // F1 Short Load Global
    SLDO_11,    // F2 Short Load Global
    SLDO_12,    // F3 Short Load Global
    SLDO_13,    // F4 Short Load Global
    SLDO_14,    // F5 Short Load Global
    SLDO_15,    // F6 Short Load Global
    SLDO_16,    // F7 Short Load Global
    SIND_0,     // F8 Short Load Indirect
    SIND_1,     // F9 Short Load Indirect
    SIND_2,     // FA Short Load Indirect
    SIND_3,     // FB Short Load Indirect
    SIND_4,     // FC Short Load Indirect
    SIND_5,     // FD Short Load Indirect
    SIND_6,     // FE Short Load Indirect
    SIND_7      // FF Short Load Indirect
};

enum pcode_csp_t
{
    CSP_IOCHECK,    // IoCheck()
    CSP_NEW,        // new()
    CSP_MOVELEFT,   // moveleft()
    CSP_MOVERIGHT,  // moveright()
    CSP_EXIT,       // exit()
    CSP_UNITREAD,   // unitread()
    CSP_UNITWRITE,  // unitwrite()
    CSP_IDSEARCH,   // idsearch()
    CSP_TREESEARCH, // treesearch()
    CSP_TIME,       // time()
    CSP_FILLCHAR,   // fillchar()
    CSP_SCAN,       // scan()
    CSP_UNITSTATUS, // unitstatus()
    CSP_LOADSEGMENT = 21, // LoadSegment()
    CSP_UNLOADSEGMENT, // UnloadSegment()
    CSP_TRUNC,      // trunc()
    CSP_ROUND,      // round()
    CSP_SIN,        // sin()
    CSP_COS,        // cos()
    CSP_TAN,        // tan()
    CSP_ATAN,       // atan()
    CSP_LN,         // ln()
    CSP_EXP,        // exp()
    CSP_SQRT,       // sqrt()
    CSP_MARK,       // mark()
    CSP_RELEASE,    // release()
    CSP_IORESULT,   // ioresult()
    CSP_UNITBUSY,   // unitbusy()
    CSP_PWROFTEN,   // PwrOfTen()
    CSP_UNITWAIT,   // unitwait()
    CSP_UNITCLEAR,  // unitclear()
    CSP_HALT,       // halt()
    CSP_MEMAVAIL    // memavail()
};

enum
{
    CXP_0_EXECERROR = 2,
    CXP_0_FINIT,
    CXP_0_FRESET,
    CXP_0_FOPEN,
    CXP_0_FCLOSE,
    CXP_0_FGET,
    CXP_0_FPUT,
    CXP_0_XSEEK,
    CXP_0_FEOF,
    CXP_0_FEOLN,
    CXP_0_FREADINT,
    CXP_0_FWRITEINT,
    CXP_0_XREADREAL,
    CXP_0_XWRITEREAL,
    CXP_0_FREADCHAR,
    CXP_0_FWRITECHAR,
    CXP_0_FREADSTRING,
    CXP_0_FWRITESTRING,
    CXP_0_FWRITEBYTES,
    CXP_0_FREADLN,
    CXP_0_FWRITELN,
    CXP_0_SCONCAT,
    CXP_0_SINSERT,
    CXP_0_SCOPY,
    CXP_0_SDELETE,
    CXP_0_SPOS,
    CXP_0_FBLOCKIO,
    CXP_0_FGOTOXY
};

enum
{
    CXP_31_FSEEK = 2,
    CXP_31_FREADREAL,
    CXP_31_FWRITEREAL,
};

enum
{
    OP_CMP_REAL = 2,
    OP_CMP_STRING = 4,
    OP_CMP_BOOLEAN = 6,
    OP_CMP_SET = 8,
    OP_CMP_BYTE_ARRAY = 10,
    OP_CMP_MULTI_WORD = 12
};

// closetype
enum
{
    CNORMAL,
    CLOCK,
    CPURGE,
    CCRUNCH
};

//
// The DECOPS procedure (seg 30, proc 4) has a selector pushed onto the
// stack immediately before the call.
//
enum
{
    DECOPS_SEG_NUM = 30,
    DECOPS_PROC_NUM = 4,

    DECOPS_ADJUST = 0,
    DECOPS_ADD = 2,
    DECOPS_SUBTRACT = 4,
    DECOPS_NEGATE = 6,
    DECOPS_MULTIPLY = 8,
    DECOPS_DIVIDE = 10,
    DECOPS_STR = 12,
    DECOPS_CONVERT_TOS_M1 = 14,
    DECOPS_COMPARE = 16,
    DECOPS_CONVERT_TOS = 18,
    DECOPS_TRUNC = 20,

    // These selectors do not exist (yet).
    DECOPS_ABS = 22,
    DECOPS_SQR = 24,
    DECOPS_MOD = 26,

    DECOPS_COMPARE_LT = 8,
    DECOPS_COMPARE_LE = 9,
    DECOPS_COMPARE_GE = 10,
    DECOPS_COMPARE_GT = 11,
    DECOPS_COMPARE_NE = 12,
    DECOPS_COMPARE_EQ = 13,

    // procedure numbers
    DECOPS_FREADDEC = 2,
    DECOPS_FWRITEDEC = 3,
};

#endif // LIB_PCODE_H
// vim: set ts=8 sw=4 et :
